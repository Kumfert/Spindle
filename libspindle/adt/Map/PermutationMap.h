//
//  PermutationMap.h -- an abstraction of a permutation vector.
//
//  $Id: PermutationMap.h,v 1.2 2000/02/18 01:31:49 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  The PermutationMap class is meant to abstract away the concept of
//  a permutation vector which may be of old2New or new2old varieties.
//
//  A PermutationMap is not in a valid state until validate() is called
//  to insure that both old2New and new2Old are consistent with each
//  other.
//

#ifndef SPINDLE_PERMUTATION_H_
#define SPINDLE_PERMUTATION_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

#ifndef SPINDLE_PERSISTANT_H_
#include "spindle/SpindlePersistant.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class PermutationMap : public SpindlePersistant {
  SPINDLE_DECLARE_PERSISTANT( PermutationMap )

private:
  PermutationMap();
  SharedArray<int> old2New;
  SharedArray<int> new2Old;

  enum { NEITHER, OLD2NEW, NEW2OLD } changedMostRecent;
  int sz;
public:
  // constructors/destructors
  PermutationMap( const int n );
  PermutationMap( const PermutationMap& perm );
  virtual ~PermutationMap();

  int size() const;

  // access specific permuation vectors
  const SharedArray<int>& getOld2New() const;
  SharedArray<int>& getOld2New();

  const SharedArray<int>& getNew2Old() const;
  SharedArray<int>& getNew2Old();


  virtual void validate();
  virtual void dump( FILE * stream ) const;

  virtual bool reset();
  void resize( const int n );
  void prettyPrint( FILE * fp, bool new2old_perm=true, bool printLength=false, 
		    bool onesOffset=false ) const; 

};

inline 
int PermutationMap::size() const {
  return sz;
}

inline 
PermutationMap::PermutationMap( const int n )
  : old2New(n), new2Old(n), sz(n) {
    incrementInstanceCount( PermutationMap::MetaData );
    reset();
}

inline 
PermutationMap::PermutationMap( const PermutationMap& perm ) 
  : sz( perm.size() ) {
    incrementInstanceCount( PermutationMap::MetaData );
    old2New.import( perm.old2New );
    new2Old.import( perm.new2Old );
    changedMostRecent = perm.changedMostRecent;
    currentState = perm.currentState;
}

inline 
PermutationMap::~PermutationMap() {
  decrementInstanceCount( PermutationMap::MetaData );
}

SPINDLE_END_NAMESPACE

#endif


