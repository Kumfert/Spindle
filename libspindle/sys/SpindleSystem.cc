//
// SpindleSystem.cc
//
// $Id: SpindleSystem.cc,v 1.2 2000/02/18 01:32:01 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//

#include "spindle/SpindleSystem.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

static int count = 0;

static FILE* TraceFile = stderr;
static FILE* WarnFile = stderr;
static FILE* ErrorFile = stderr;

static unsigned int TraceLevel = 0;
static PersistanceRegistry * Registry = 0;
static StackTrace * Stack = 0;

static char* currentFunction;
static char* currentFile;
static int currentLine;

PersistanceRegistry& 
SpindleSystem::registry() {
  return *Registry;
}

StackTrace& 
SpindleSystem::stack() {
  return *Stack;
}

void
SpindleSystem::startSystem() {
  if (count++ == 0) {
    Registry = new PersistanceRegistry();
    Stack = new StackTrace();
  }
}

void
SpindleSystem::stopSystem() {
  if ( --count==0 ) {
    delete Registry;
    delete Stack;
  }
}



void
SpindleSystem::setWhere( char* func, char* file, int line ) {
  currentFunction = func;
  currentFile = file;
  currentLine = line;
}

void 
SpindleSystem::setTraceFile( FILE* traceFile ) {
  TraceFile = traceFile;
}

void 
SpindleSystem::setWarnFile( FILE* warnFile ) {
  WarnFile = warnFile;
}


void 
SpindleSystem::setErrorFile( FILE* errorFile ) {
  ErrorFile = errorFile;
}

void 
SpindleSystem::setTraceLevel( unsigned int traceLevel ) {
  if ( traceLevel >= 0 ) {
    TraceLevel = traceLevel;
  }
}

unsigned int 
SpindleSystem::getTraceLevel() { 
  return TraceLevel;
}

//
// Now implement the print functions for standard libraries.
//
//
#if SPINDLE_TARGET == std

void 
SpindleSystem::assert_that( bool arg, const char* format, ... ) {
  if ( (! arg)  && (ErrorFile != 0 ) ) {
    va_list ap;
    va_start( ap, format);
    fprintf(ErrorFile, "\nASSERT FAILED:  ");
    vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( ErrorFile );
    SpindleSystem::stack().reset();
#else
    SpindleSystem::printWhere( ErrorFile );
#endif
    va_end(ap);
    // exit(-1);
  }
}

void 
SpindleSystem::error( unsigned long errorCode, const char* format, ... ) {
  va_list ap;
  va_start( ap, format );
  if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%ld):  ",errorCode);
    vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( ErrorFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

merr
SpindleSystem::mError(  int line, char* file, char* func, merr errcode, char* msg ) {
  if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
  }
  return errcode;
}

void 
SpindleSystem::warn_if( bool arg, const char* format, ... ) {
  if ( arg && ( WarnFile != 0 ) ) {
    va_list ap;
    va_start( ap, format);
    fprintf(WarnFile, "\nWARNING:  ");
    vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( WarnFile );
#else 
    SpindleSystem::printWhere( WarnFile );
#endif
    va_end(ap);
  }
}

void 
SpindleSystem::warning( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( WarnFile != 0 ) {
    fprintf(WarnFile, "\nWARNING:  ");
    vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( WarnFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

void 
SpindleSystem::trace( unsigned int traceLevel, const char* format, ... ) {
  if ( traceLevel & TraceLevel ) {
    va_list ap;
    va_start( ap, format);
    //    const StackInfo& si = SpindleSystem::stack().top();
    //    if ( traceLevel != SPINDLE_TRACE_STACK ) {
    //      fprintf(TraceFile, "\nTrace(%d) %s [%d] - \"%s\"\n\t",traceLevel, si.file, si.line, si.function );
    //    }
    if ( TraceFile != 0 ) {
      vfprintf(TraceFile, format, ap);
    }
    va_end(ap);
  }
}

void 
SpindleSystem::printf( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( TraceFile != 0 ) {
    vfprintf(TraceFile, format, ap);
  }
  va_end(ap);
}

void 
SpindleSystem::printWhere( FILE * fp ) {
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp != 0 ) ) {
    fprintf(fp, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
}

//
// else make a library using matlab constructs...
//
//
#elif SPINDLE_TARGET == matlab 

extern "C" {
#include "mex.h"
}
static char buffer[512*8];

void 
SpindleSystem::assert_that( bool arg, const char* format, ... ) {
  if ( ! arg ) {
    va_list ap;
    va_start( ap, format);
    if ( ErrorFile == 0 )  { // print via Matlab's messaging system.
      int i = sprintf( buffer, "\nASSERT FAILED:  ");
      vsprintf( buffer+i, format, ap );
      mexWarnMsgTxt( buffer );
    } else {  // assume that a file is open
      fprintf(ErrorFile, "\nASSERT FAILED:  ");
      vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
      SpindleSystem::stack().dump( ErrorFile );
      SpindleSystem::stack().reset();
#else
      SpindleSystem::printWhere( ErrorFile );
#endif
    }
    va_end(ap);
    // exit(-1);
  }
}

void 
SpindleSystem::error( unsigned long errorCode, const char* format, ... ) {
  va_list ap;
  va_start( ap, format );
  if ( ErrorFile == 0 ) {
      int i = sprintf( buffer, "\nASSERT FAILED:  ");
      vsprintf( buffer+i, format, ap );
      mexWarnMsgTxt( buffer );
  } else if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%ld):  ",errorCode);
    vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( ErrorFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

merr
SpindleSystem::mError(  int line, char* file, char* func, merr errcode, char* msg ) {
  if ( ErrorFile == 0 ) {
      vsprintf( buffer, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
      mexWarnMsgTxt( buffer );
  } else if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
  }
  return errcode;
}

void 
SpindleSystem::warn_if( bool arg, const char* format, ... ) {
  if ( arg ) {
    va_list ap;
    va_start( ap, format);
    if ( WarnFile == 0 ) {
      int i = sprintf( buffer, "\nWARNING:   ");
      vsprintf( buffer+i, format, ap );
      mexWarnMsgTxt( buffer );
    } else {
      fprintf(WarnFile, "\nWARNING:  ");
      vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
      SpindleSystem::stack().dump( WarnFile );
#else 
      SpindleSystem::printWhere( WarnFile );
#endif
    }
    va_end(ap);
  }
}

void 
SpindleSystem::warning( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( WarnFile == 0 ) {
    int i = sprintf( buffer, "\nWARNING:  ");
    vsprintf( buffer+i, format, ap );
    mexWarnMsgTxt( buffer );
  } else if ( WarnFile != 0 ) {
    fprintf(WarnFile, "\nWARNING:  ");
    vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( WarnFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

void 
SpindleSystem::trace( unsigned int traceLevel, const char* format, ... ) {
  if ( traceLevel & TraceLevel ) {
    va_list ap;
    va_start( ap, format);
    //    const StackInfo& si = SpindleSystem::stack().top();
    //    if ( traceLevel != SPINDLE_TRACE_STACK ) {
    //      fprintf(TraceFile, "\nTrace(%d) %s [%d] - \"%s\"\n\t",traceLevel, si.file, si.line, si.function );
    //    }
    if ( TraceFile == 0 ) {
      vsprintf( buffer, format, ap );
      mexPrintf( "%s", buffer );
    } else if ( TraceFile != 0 ) {
      vfprintf(TraceFile, format, ap);
    }
    va_end(ap);
  }
}

void 
SpindleSystem::printf( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( TraceFile == 0 ) {
    vsprintf( buffer, format, ap );
    mexPrintf( "%s", buffer );
  } else  if ( TraceFile != 0 ) {
    vfprintf(TraceFile, format, ap);
  }
  va_end(ap);
}

void 
SpindleSystem::printWhere( FILE * fp ) {
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp != 0 ) ) {
    fprintf(fp, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp == 0 ) ) {
    sprintf( buffer, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
    mexPrintf( "%s", buffer );
  }
}

#elif  SPINDLE_TARGET == petsc 

extern "C" {
#include "petsc.h"
}
static char buffer[512*8];

void 
SpindleSystem::assert_that( bool arg, const char* format, ... ) {
  if ( ! arg  && ( ErrorFile != 0 ) ) {
    va_list ap;
    va_start( ap, format);
    int i = sprintf( buffer, "\nASSERT FAILED:  ");
    vsprintf( buffer+i, format, ap);
    PetscFPrintf( PETSC_COMM_SELF, ErrorFile, "%s", buffer );
    SpindleSystem::printWhere( ErrorFile );
    va_end(ap);
    // exit(-1);
  }
}

void 
SpindleSystem::error( unsigned long errorCode, const char* format, ... ) {
  va_list ap;
  va_start( ap, format );
  if ( ErrorFile != 0 ) {
    int i = sprintf( buffer, "\nERROR(%ld):  ", errorCode );
    vsprintf( buffer+i, format, ap);
    PetscFPrintf( PETSC_COMM_SELF, ErrorFile, "%s", buffer );
    SpindleSystem::printWhere( ErrorFile );
    printWhere( ErrorFile );
  }
  va_end(ap);
}

merr
SpindleSystem::mError(  int line, char* file, char* func, merr errcode, char* msg ) {
  if ( ErrorFile != 0 ) {
    PetscFPrintf( PETSC_COMM_SELF, ErrorFile, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
  }
  return errcode;
}

void 
SpindleSystem::warn_if( bool arg, const char* format, ... ) {
  if ( arg ) {
    va_list ap;
    va_start( ap, format);
    if ( WarnFile != 0 ) {
      int i = sprintf( buffer, "\nWARNING:  ");
      vsprintf( buffer+i, format, ap );
      PetscFPrintf( PETSC_COMM_SELF, WarnFile, "%s", buffer );
      SpindleSystem::printWhere( WarnFile );
    }
    va_end(ap);
  }
}

void 
SpindleSystem::warning( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( WarnFile != 0 ) {
    int i = sprintf( buffer, "\nWARNING:  ");
    vsprintf(buffer+i, WarnFile, format, ap);
    PetscFPrintf( PETSC_COMM_SELF, WarnFile, "%s", buffer );
    printWhere( WarnFile );
  }
  va_end(ap);
}

void 
SpindleSystem::trace( unsigned int traceLevel, const char* format, ... ) {
  if ( traceLevel & TraceLevel ) {
    va_list ap;
    va_start( ap, format);
    //    const StackInfo& si = SpindleSystem::stack().top();
    //    if ( traceLevel != SPINDLE_TRACE_STACK ) {
    //      fprintf(TraceFile, "\nTrace(%d) %s [%d] - \"%s\"\n\t",traceLevel, si.file, si.line, si.function );
    //    }
    if ( TraceFile != 0 ) {
      vsprintf(buffer , format, ap);
      PetscFPrintf( PETSC_COMM_SELF, TraceFile, "%s", buffer );
    }
    va_end(ap);
  }
}

void 
SpindleSystem::printf( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  vsprintf( buffer , format, ap );
  PetscPrintf( PETSC_COMM_SELF,  "%s", buffer );
  va_end(ap);
}

void 
SpindleSystem::printWhere( FILE * fp ) {
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp != 0 ) ) {
    PetscFPrintf( PETSC_COMM_SELF, fp, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp == 0 ) ) {
    PetscPrintf( PETSC_COMM_SELF, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
}

#else
#error Error do not recognize SPINDLE TARGET of SPINDLE_TARGET
#endif
