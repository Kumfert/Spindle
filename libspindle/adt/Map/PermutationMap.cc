//
// PermutationMap.cc
//
//  $Id: PermutationMap.cc,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is povided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  The PermutationMap class is meant to abstract away the concept of
//  a permutation vector which may be of old2New or new2old varieties.
//
//  A PermutationMap is not in a valid state until finalize() is called
//  to insure that both old2New and new2Old are consistent with each
//  other.
//

#include "spindle/PermutationMap.h"

#ifndef SPINDLE_ARCHIVE_H_
#include "spindle/SpindleArchive.h"
#endif

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

#ifdef __FUNC__
#undef __FUNC__
#endif

PermutationMap::PermutationMap() { 
  incrementInstanceCount( PermutationMap::MetaData );
  currentState = EMPTY;
}

void
PermutationMap::validate() {
  if ( currentState == VALID ) { return; }
  int *src; 
  int *dest;
  if ( changedMostRecent == NEITHER ) {
    src = old2New.begin();
    dest = new2Old.begin();
    if ((*src >= 0)&&(*dest <= 0)) { 
      changedMostRecent = OLD2NEW; 
    } else if ((*src <= 0)&&(*dest >= 0)) { 
      changedMostRecent = NEW2OLD; 
      src = new2Old.begin();
      dest = old2New.begin();
    }
  } else if ( changedMostRecent == OLD2NEW ) { 
    src = old2New.begin();
    new2Old.init( 0 );
    dest = new2Old.begin();
  } else { // if ( changedMostRecent == NEW2OLD ) {
    src = new2Old.begin();
    old2New.init( 0 );
    dest = old2New.begin();
  }
  {for (int i=0; i<sz; i++) { 
    int perm_i = src[i];
    if ( perm_i == -1 ) { continue; }
    if ( ( perm_i < 0) ||  ( perm_i >= sz ) ) { 
      currentState = INVALID;
      return; 
    }
    dest[perm_i]++;
  }}
  // ... now make sure each vertex only occurred once
  int * p = dest;
  { for (int i=0; i<sz; i++) { 
    if ( ( *p != 1 ) && (*p != 0) ) { 
      currentState = INVALID;
      return;
    }
  }}

  // ... finally copy the reverse permutation 
  {for (int i=0; i<sz; i++) { 
    dest[ src[ i ] ] = i; 
  }}
  currentState = VALID;
  changedMostRecent = NEITHER;
  return;
}

bool
PermutationMap::reset() {
  old2New.init( -1 );
  new2Old.init( -1 );
  currentState = EMPTY;
  changedMostRecent = NEITHER;
  return true;
}


const SharedArray<int>& 
PermutationMap::getOld2New() const {
  return old2New;
}

SharedArray<int>& 
PermutationMap::getOld2New() {
  if ( (!isValid() ) && (changedMostRecent == NEW2OLD) ) {
    validate();
    assert( isValid() );
  }
  currentState = UNKNOWN;
  changedMostRecent = OLD2NEW;
  return old2New;
}

const SharedArray<int>& 
PermutationMap::getNew2Old() const {
  return new2Old;
}

SharedArray<int>& 
PermutationMap::getNew2Old() {
  if ((!isValid()) && (changedMostRecent == OLD2NEW) ) {
    validate();
    assert( isValid() );
  }
  currentState = UNKNOWN;
  changedMostRecent = NEW2OLD;
  return new2Old;
}

void 
PermutationMap::resize( const int n ) { 
  currentState = EMPTY;
  changedMostRecent = NEITHER;
  sz = n;
  old2New.resize(n);
  new2Old.resize(n);
}

void
PermutationMap::dump( FILE * fp ) const {
  if ( isValid() ) {
    fprintf(fp,"PermutationMap:");
    const int* p = getNew2Old().lend();
    for( int i=0; i<size(); i++) {
      fprintf(fp,"%d ", p[i] );
    }
  }
}

void
PermutationMap::prettyPrint( FILE * fp, bool new2old_perm, bool printLength, 
			     bool onesOffset ) const { 

  if ( !isValid() ) {
    return ;
  }
  int offset = (onesOffset) ? 1 : 0 ;
  
  if ( printLength == true ) {
    fprintf( fp,"%d\n", size() );
  }
  
  const int * p;
  if ( new2old_perm ) { 
    p = new2Old.lend();
  } else { 
    p = old2New.lend();
  }

  for(const int* cur = p, *stop=p + size(); cur <stop; ++cur ) {
    fprintf( fp, "%d\n", *cur + offset);
  }
} 

void
PermutationMap::storeObject( SpindleArchive& ar ) const {
  if ( isValid() ) {
    const int *p = getNew2Old().lend();
    size_t my_size = size();
    pair< const int*, size_t> my_pair( p,my_size );
    ar << my_pair;
  } else {
    ar << -1;
  }
}
#define __FUNC__ "PermutationMap::loadObject( SpindleArchive& ar )"
void
PermutationMap::loadObject( SpindleArchive& ar ) {
  size_t new_size = ar.peekLength();
  if ( (int)new_size == -1 ) { 
    int junk;
    ar >> junk;
    currentState = INVALID;
  } else {
    resize(new_size);
    int* p = getNew2Old().begin();
    pair<int*, size_t> new_pair( p, new_size );
    
    ar >> new_pair;
    
    WARN_IF( new_pair.second != new_size, "Warning, expected size=%d, but extracted only %d.", 
	     new_pair.second, new_size );
    changedMostRecent = NEW2OLD;
    validate();
  }
}
#undef __FUNC__

SPINDLE_IMPLEMENT_PERSISTANT( PermutationMap, SpindlePersistant )


