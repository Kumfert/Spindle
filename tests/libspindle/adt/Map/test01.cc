//
// src/adt/Map/drivers/tests/test01.cc -- tests PermutationMap
//
//  $Id: test01.cc,v 1.2 2000/02/18 01:32:07 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is povided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//


#include "spindle/PermutationMap.h"

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#define REGISTRY SpindleSystem::registry()

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

int main() {
  SpindleSystem::registry().setDumpOnDestroy( stdout );
  SpindleSystem::setErrorFile( stdout );

  int perm_array[] = { 3, 4, 2, 0, 1};
  PermutationMap perm(5);
  perm.getNew2Old().take( perm_array, 5 );
  perm.validate();


  cout << "Size of PermutationMap class = " << sizeof(PermutationMap) << endl;
  
  const ClassMetaData* metaData = REGISTRY.findClass("PermutationMap");
  
  cout << "According to the registry it is = " << metaData->classSize << endl;
  cout << "This perm object is serial #" << perm.queryIDNumber() << endl;

  PermutationMap perm2(8);
  cout << "The second PermutationMap object has serial #" << perm2.queryIDNumber() << endl;

}
