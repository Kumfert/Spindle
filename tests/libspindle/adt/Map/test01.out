Size of PermutationMap class = 44
According to the registry it is = 44
This perm object is serial #0
The second PermutationMap object has serial #1
Destroying Registry... PersistanceRegistery::dump():  3 items
             PermutationMap (44) 2/2---> SpindlePersistant
           SpindleBaseClass (8) 2/2---> 
          SpindlePersistant (12) 2/2---> SpindleBaseClass
